src/GafferOSL/OSLCode.cpp

```diff
@@ -0,0 +1,416 @@
+
+#include "GafferOSL/Private/CapturingErrorHandler.h"
+#include "GafferOSL/OSLCode.h"
+
+using namespace std;
+using namespace IECore;
+using namespace OSL;
+using namespace Gaffer;
+using namespace GafferOSL;
+
+//////////////////////////////////////////////////////////////////////////
+// Code generation
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+string colorSplineParameter( const SplinefColor3fPlug *plug )
+{
+	string result;
+	result += "\tfloat " + plug->getName().string() + "Positions[] = { 0, 0, 1, 1 },\n";
+	result += "\tcolor " + plug->getName().string() + "Values[] = { 0, 0, 1, 1 },\n";
+	result += "\tstring " + plug->getName().string() + "Basis = \"catmull-rom\",\n";
+	return result;
+}
+
+string parameter( const Plug *plug )
+{
+	const Gaffer::TypeId plugType = (Gaffer::TypeId)plug->typeId();
+	if( plugType == SplinefColor3fPlugTypeId )
+	{
+		return colorSplineParameter( static_cast<const SplinefColor3fPlug *>( plug ) );
+	}
+
+	string type;
+	string defaultValue;
+	switch( plugType )
+	{
+		case FloatPlugTypeId :
+			defaultValue = "0.0";
+			type = "float";
+			break;
+		case IntPlugTypeId :
+			defaultValue = "0";
+			type = "int";
+			break;
+		case Color3fPlugTypeId :
+			defaultValue = "color( 0.0 )";
+			type = "color";
+			break;
+		case V3fPlugTypeId :
+			defaultValue = "vector( 0.0 )";
+			type = "vector";
+			break;
+		case M44fPlugTypeId :
+			defaultValue = "1";
+			type = "matrix";
+			break;
+		case StringPlugTypeId :
+			defaultValue = "\"\"";
+			type = "string";
+			break;
+		case PlugTypeId :
+			defaultValue = "0";
+			type = "closure color";
+			break;
+		default :
+			throw IECore::Exception( string( "Unsupported plug type \"" ) + plug->typeName() + "\"" );
+	}
+
+	string direction = plug->direction() == Plug::Out ? "output " : "";
+	return "\t" + direction + type + " " + plug->getName().string() + " = " + defaultValue + ",\n";
+}
+
+string generate( const OSLCode *shader, string &shaderName )
+{
+	// Start with parameters
+
+	string result;
+
+	for( PlugIterator it( shader->parametersPlug() ); !it.done(); ++it )
+	{
+		result += parameter( it->get() );
+	}
+
+	result += "\n";
+
+	for( PlugIterator it( shader->outPlug() ); !it.done(); ++it )
+	{
+		result += parameter( it->get() );
+	}
+
+	result += ")\n";
+
+	// Add on body
+
+	const std::string code = shader->codePlug()->getValue();
+	result += "{\n" + code;
+
+	if( code.size() && *code.rbegin() != ';' )
+	{
+		result += ";";
+	}
+
+	result += "\n}\n";
+
+	// The result so far uniquely characterises our shader,
+	// so generate the shader name from its hash.
+
+	IECore::MurmurHash hash;
+	hash.append( result );
+	shaderName = "oslCode" + hash.toString();
+
+	// Now we have the shader name, we can add on the header
+	// for the shader, consisting of a constant set of includes
+	// and the shader name itself.
+
+	string header;
+	header += "#include \"GafferOSL/ObjectProcessing.h\"\n";
```

Comment:
Yeah, I think it'd be cool too. I must admit I avoided doing it partly out of laziness, but also because the necessary regexes might be fragile, and not be kept up to date with changes in the headers. I figured the lazy route is safer, and not too big a deal for anyone savvy enough to decide they want to work on the exported shader.